File: Plugins/ProductionCapacity/Lemoine.Plugin.ProductionCapacityPerOperation/ProductionCapacityExtension.fs
(*
 * Copyright (C) 2009-2023 Lemoine Automation Technologies
 *
 * SPDX-License-Identifier: Apache-2.0
 *)

namespace Lemoine.Plugin.ProductionCapacityPerOperation

open System.Collections.Generic

open Lemoine.Core.Log
open Lemoine.Core.Cache
open Lemoine.Extensions.Business.Cache
open Lemoine.Model
open Lemoine.ModelDAO

type ProductionCapacityExtension () =
  inherit Lemoine.Extensions.MultipleInstanceConfigurableExtension<Configuration>()

  let log = LogManager.GetLogger ("Lemoine.Plugin.ProductionCapacityPerOperation.ProductionCapacityExtension")

  let mutable operation: IOperation = null
  let intermediateWorkPieces = new List<IIntermediateWorkPiece> ()
  let mutable configuration: Configuration = new Configuration ()

  interface Lemoine.Extensions.IExtension with
    member this.UniqueInstance = this.UniqueInstance

  interface Lemoine.Extensions.Business.Operation.IProductionCapacityExtension with
    member this.Initialize(machine) =
      match this.LoadConfiguration (&configuration) with
      | false -> false
      | true ->
        begin
          use session = ModelDAOHelper.DAOFactory.OpenSession ()
          use transaction = session.BeginReadOnlyTransaction ("Plugin.ProductionCapacityPerOperation.ProductionCapacityExtension")
          let operation = ModelDAOHelper.DAOFactory.OperationDAO.FindById configuration.OperationId in
          if isNull operation
          then log.Error (sprintf "Operation with id=%d does not exist" configuration.OperationId) |> fun () -> false
          else
            begin
              for intermediateWorkPiece in operation.IntermediateWorkPieces do
                intermediateWorkPieces.Add intermediateWorkPiece;
              true
            end
        end

    member this.GetCapacityPerHourAsync(intermediateWorkPiece: IIntermediateWorkPiece): System.Threading.Tasks.Task<float> = 
      task {
        if Seq.exists (fun (iwp: IIntermediateWorkPiece) -> iwp.Id = intermediateWorkPiece.Id) intermediateWorkPieces
        then return configuration.CapacityPerHour
        else return raise <| new System.Exception "Not supported intermediate work piece"
      }

    member this.Score = configuration.Score

File: Plugins/ProductionCapacity/Lemoine.Plugin.ProductionCapacityPerMachine/ProductionCapacityExtension.cs
// Copyright (C) 2009-2023 Lemoine Automation Technologies
//
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Lemoine.Core.Log;
using Lemoine.Extensions.Business.Operation;
using Lemoine.Model;
using Lemoine.ModelDAO;

namespace Lemoine.Plugin.ProductionCapacityPerMachine
{
  public class ProductionCapacityExtension
    : Lemoine.Extensions.MultipleInstanceConfigurableExtension<Configuration>
    , IProductionCapacityExtension
  {
    readonly ILog log = LogManager.GetLogger (typeof (ProductionCapacityExtension).FullName);

    IMachine m_machine;
    Configuration m_configuration;

    public double Score => m_configuration.Score;

    public bool Initialize (IMachine machine)
    {
      m_machine = machine;

      if (!LoadConfiguration (out m_configuration)) {
        log.Warn ("Initialize: the configuration is not valid");
        return false;
      }

      if (!(m_configuration.MachineIds is null) && m_configuration.MachineIds.Any ()) {
        if (!m_configuration.MachineIds.Contains (machine.Id)) {
          if (log.IsDebugEnabled) {
            log.Debug ($"Initialize: the machine does not match => return false");
          }
          return false;
        }
      }

      if (0 != m_configuration.MachineFilterId) {
        using (var session = ModelDAOHelper.DAOFactory.OpenSession ())
        using (var transaction = session.BeginReadOnlyTransaction ("Lemoine.Plugin.ProductionCapacityPerMachine.Initialize")) {
          var machineFilter = ModelDAOHelper.DAOFactory.MachineFilterDAO
            .FindById (m_configuration.MachineFilterId);
          if (machineFilter is null) {
            log.Error ($"Initialize: machine filter with id={m_configuration.MachineFilterId} does not exist => return false");
            return false;
          }
          if (!machineFilter.IsMatch (machine)) {
            if (log.IsDebugEnabled) {
              log.Debug ($"Initialize: machine filter does not match");
            }
            return false;
          }
        }
      }

      return true;
    }

    public Task<double> GetCapacityPerHourAsync (IIntermediateWorkPiece intermediateWorkPiece)
    {
      var result = m_configuration.CapacityPerHour;
      return Task.FromResult (result);
    }

  }
}
